### PR TITLE
Feature/61 json backup

### PR DESCRIPTION
**Pull request by cayop**
Now a json file is always saved with the snapshot that is sent to the devicehub. If all process it's ok then this file is delete.
The directory where the snapshots reside is "tmp/snapshots"
The json file has this format: "uuid_unixtime.json"
There are a new test file in basic.snapshot.badly_formed.yaml with errors
In tests/test_snapshot.py there are test_backup_snapshot_with_errors. This test is very useful for check and adapt different situations

## Requirements definition
- Error (unprocesable entity: 422).
- Definir JSON name to **know which client has made the request**: 
   **Proposal**: snapshot + username + date (y-m-d).json

## TODO
- [ ] add client username to JSON filename.

### CODE REVIEW

File: tests/test_snapshot.py [hunk 1]
```diff
@@ -473,3 +478,48 @@ def test_pc_rating_rate_none(user: UserClient):
...
+    """ This test check if works the function save_snapshot_in_file """
+    snapshot_no_hid = file('basic.snapshot.nohid')
+    save_json(snapshot_no_hid)
+
+    uuid = snapshot_no_hid['uuid']
+    files = [x for x in os.listdir(TMP_SNAPSHOTS) if uuid in x]
+
+    snapshot = {'software': '', 'version': '', 'uuid': ''}
+    if files:
+        path_snapshot = "{dir}/{file}".format(dir=TMP_SNAPSHOTS, file=files[0])
```
Comment (slamora) [inline tests/test_snapshot.py:494]: Could you replace format string with `os.path.join` to generate full path?

File: tests/test_snapshot.py [hunk 2]
```diff
@@ -473,3 +478,48 @@ def test_pc_rating_rate_none(user: UserClient):
...
+        file_snapshot = open(path_snapshot)
+        snapshot = json.loads(file_snapshot.read())
+        file_snapshot.close()
```
Comment (slamora) [inline tests/test_snapshot.py:497]: Maybe this 3 lines could be refactored using `with open(....):` [with statement](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files). That automatically closes the file.

File: tests/test_snapshot.py [hunk 3]
```diff
@@ -473,3 +478,48 @@ def test_pc_rating_rate_none(user: UserClient):
...
+    snapshot = {'software': '', 'version': '', 'uuid': ''}
+    with pytest.raises(KeyError):
+        response = user.post(res=Snapshot, data=snapshot_no_hid)
+
+    files = [x for x in os.listdir(TMP_SNAPSHOTS) if uuid in x]
+    if files:
+        path_snapshot = "{dir}/{file}".format(dir=TMP_SNAPSHOTS, file=files[0])
+        file_snapshot = open(path_snapshot)
+        snapshot = json.loads(file_snapshot.read())
+        file_snapshot.close()
```
Comment (slamora) [inline tests/test_snapshot.py:520]: Same comments as on the previous test method.

Comment (cayop) [reply tests/test_snapshot.py]: Thank you.
I put "with statement" and I used os.path.join in the test and in the view of actions.

**Comment by slamora** [comment]
@JNadeu could you describe or even better attach JSON samples of known situations which cause validation errors while processing a snapshot? This way we will be able to create more tests to cover them.

File: ereuse_devicehub/resources/action/views.py [hunk 4]
```diff
@@ -17,23 +21,46 @@
 
 SUPPORTED_WORKBENCH = StrictVersion('11.0')
 
+TMP_SNAPSHOTS = 'tmp/snapshots'
```
Comment (slamora) [inline ereuse_devicehub/resources/action/views.py:24]: Are we going to store user JSONs on a folder contained by the source code?

File: ereuse_devicehub/resources/action/views.py [hunk 5]
```diff
@@ -1,9 +1,13 @@
+import os
+import json
+from time import time
 from distutils.version import StrictVersion
 from typing import List
 from uuid import UUID
 
 from flask import current_app as app, request, g
 from sqlalchemy.util import OrderedSet
+from marshmallow.exceptions import ValidationError as mValidationError
```
Comment (slamora) [inline ereuse_devicehub/resources/action/views.py:10]: Is that an unused import?

**Review by slamora — CHANGES_REQUESTED** [review]

File: ereuse_devicehub/resources/action/views.py (hunk 5 above)
Comment (cayop) [reply ereuse_devicehub/resources/action/views.py]: * imports unused removed
* TMP_SNAPSHOT defined in config or in env file as you prefere
* email added to the name of file of json
* date more readable in the name of file of json

**Comment by JNadeu** [comment]
Most frequent known situations that fail when a snapshot is uploaded:

#### Error 500 INTERNAL SERVER ERROR

It is usually related to the action rate that It's not possible to compute cause don't have all the necessary data.
I attach snapshots example of this situations. 

1. Benchmark CPU don't exists ([failed.snapshot.500.missing-cpu-benchmark.json.txt](https://github.com/eReuse/devicehub-teal/files/5373513/failed.snapshot.500.missing-cpu-benchmark.json.txt))

{AttributeError: 'StopIteration' object has no attribute 'code'} = 'Missing benchmark cpu action inside CPU component'

2. Benchmark HDD don't exists ([failed.snapshot.500.missing-hdd-benchmark.json.txt](https://github.com/eReuse/devicehub-teal/files/5373510/failed.snapshot.500.missing-hdd-benchmark.json.txt))

{AttributeError: 'LookupError' object has no attribute 'code'} = 'Missing benchmark hdd action inside HDD component'


#### Error 422 UNPROCESSABLE ENTITY

It is usually related to the action price that It's not possible to compute cause don't have the chassis field or diferent than this [enum](https://github.com/eReuse/devicehub-teal/blob/833bf127191626362b259859d420089fade6ad38/ereuse_devicehub/resources/enums.py#L211).

1. "chassis": "Missing data for required field."  ([failed.snapshot.422.null-chassis.json.txt](https://github.com/eReuse/devicehub-teal/files/5373511/failed.snapshot.422.null-chassis.json.txt))

2. "chassis": "Invalid enum member Notebook" ([failed.snapshot.422.missing-chassis.json.txt](https://github.com/eReuse/devicehub-teal/files/5373512/failed.snapshot.422.missing-chassis.json.txt))

There may be other specific cases that are not covered above, but the most repeated are the above cases.

Actually if it would not be not necessary to compute the action Rate and Price these 4 cases would not be errors... This would be another issue but I wanted to leave the reflection in writing here.

**Comment by cayop** [comment]
Thank you. I add this files as a test.
All it's ok

File: ereuse_devicehub/resources/action/views.py [hunk 6]
```diff
@@ -21,12 +21,18 @@
...
     The file need to be saved with one name format with the stamptime and uuid joins
     """
-    name_file = "{uuid}_{time}.json".format(uuid=req_json.get('uuid', ''), time=int(time()))
+    uuid = req_json.get('uuid', '')
+    now = datetime.now()
+    year = now.year
+    month = now.month
+    day = now.day
+
+    name_file = f"{uuid}_{user}_{year}-{month}-{day}.json"
```
Comment (JNadeu) [inline ereuse_devicehub/resources/action/views.py:35]: Now I have realized that to have ordered by date when the snapshot is uploaded, the name of the json would have to have the date first, I would even add the {hour}-{minutes}.

Then I propose:
**name_file = f"{year}-{month}-{day}-{hour}-{minutes}_{user}_{uuid}.json"**

**Review by JNadeu — COMMENTED** [review]
I see everything right!  :-)

File: ereuse_devicehub/resources/action/views.py (hunk 6 above)
Comment (cayop) [reply ereuse_devicehub/resources/action/views.py]: Done

**Pull request merged**